runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -1819,21 +1885,35 @@ MM_CopyForwardScheme::discardRemainingCache(MM_EnvironmentVLHGC *env, MM_CopySca
 	if ((0 != discardSize) || (0 != wastedMemory)) {
 		Assert_MM_true((0 == wastedMemory) || (wastedMemory < cacheAlloc - (UDATA)cache->cacheBase));
 		MM_HeapRegionDescriptorVLHGC *region = (MM_HeapRegionDescriptorVLHGC*)_regionManager->tableDescriptorForAddress(cache->cacheBase);
-		MM_MemoryPoolBumpPointer *pool = (MM_MemoryPoolBumpPointer *)region->getMemoryPool();
+		MM_MemoryPoolAddressOrderedList *pool = (MM_MemoryPoolAddressOrderedList*)region->getMemoryPool();
 		cacheLock->acquire();
-		UDATA potential = (UDATA)region->getHighAddress() - (UDATA)cache->cacheAlloc;
-		if ((0 != discardSize) && (pool->getAllocationPointer() == cache->cacheTop) && (potential >= pool->getMinimumFreeEntrySize())) {
-			/* try to give this back */
-			pool->rewindAllocationPointerTo(cache->cacheAlloc);
-			discardSize = 0;
-			cache->cacheTop = cache->cacheAlloc;
-			/* also ensure that we update the cached mark map's atomic tail slot since we have changed the tail */
-			env->_copyForwardCompactGroups[cache->_compactGroup]._markMapAtomicTailSlotIndex = _markMap->getSlotIndex((J9Object *)cache->cacheTop);
-		}
-		UDATA totalFreeToReturn = discardSize + wastedMemory;
+		UDATA totalFreeToReturn = wastedMemory + discardSize;
+
+		/* TODO: recycling remanining cache, maybe can connect to one exist free memory */
+//		UDATA totalFreeToReturn = wastedMemory;
+//		PORT_ACCESS_FROM_ENVIRONMENT(env);
+//		j9tty_printf(PORTLIB, "env=%p, discardRemainingCache region=%p, wastedMemory=%zu, discardSize=%zu\n", env, region, wastedMemory, discardSize);
+//
+//		if (pool->coalesceHeapChunk(env, (void*)cacheAlloc, (void*)cacheTop)) {
+//			j9tty_printf(PORTLIB, "env=%p, discardRemainingCache coalesceHeapChunk cacheAlloc=%p, cacheTop=%p\n", env, cacheAlloc, cacheTop);
+//			cache->cacheTop = cache->cacheAlloc;
+//			/* also ensure that we update the cached mark map's atomic tail slot since we have changed the tail */
+//			env->_copyForwardCompactGroups[cache->_compactGroup]._markMapAtomicTailSlotIndex = _markMap->getSlotIndex((J9Object *)cache->cacheTop);
+////		} else if (!pool->recycleHeapChunk(env, (void*)cacheAlloc, (void*)cacheTop)) {
+//		} else if (!pool->recycleHeapChunk((void*)cacheAlloc, (void*)cacheTop)) {
+//			totalFreeToReturn += discardSize;
+//		} else {
+//			j9tty_printf(PORTLIB, "env=%p, discardRemainingCache recycleHeapChunk cacheAlloc=%p, cacheTop=%p\n", env, cacheAlloc, cacheTop);
+//		}
 		if (0 != totalFreeToReturn) {
-			pool->setFreeMemorySize(pool->getActualFreeMemorySize() + totalFreeToReturn);
+			pool->incrementDarkMatterBytes(totalFreeToReturn);
```

Comment:
While this is ok for this round of changes, we should eventually try do better. It's not exactly straightforward whether we should try to build thread local list of remainders (per compactGroup) or something else, but again, I don't want to hold these changes for that.